[RFC 0125]: Use an intermediary representation (bootspec) for bootloader backends (#125)
NixOS/rfcs: rfcs/0125-bootspec.md

---
feature: bootspec
start-date: 2022-05-09
author: Graham Christensen
co-authors: Cole Helbling
shepherd-team: (names, to be nominated and accepted by RFC steering committee)
shepherd-leader: (name to be appointed by RFC steering committee)

> related-issues: (will contain links to implementation PRs)

Author: ```suggestion
related-issues: https://github.com/NixOS/nixpkgs/pull/172237
```

This is a valid method of implementation, but is insufficient for larger deployments.

An enterprise deployment of SecureBoot may have a centralized signing service with careful auditing about what is signed.
It is more likely that the signed bootloader files come prebuilt and presigned from upstream and an unsigned file means a violation of policy.
This is also a valid implementation, but is too heavy for a single user.

There are infinitely many policy choices and implementations possible, and a good solution here is to allow deployers of NixOS to implement their own bootloader management tools.
By creating a well defined specification of generations and the boot-relevant data we enable this external development.

## What about systemd's Bootloader Specification?
[systemd-bootloader-specification]: #systemd-bootloader-specification

Systemd's bootloader specification is a good format for a different problem.
A single NixOS generation can contain multiple bootable systems and options, with additional features unique to NixOS built on top.
Most Linux distributions don't deal with many unique and ever-changing bootables.

> This proposal is specifically to deal with the collection of bootables and improve our ability to interface with.

Reviewer: interface with what?

Author: I'm not totally sure what I was thinking there. Perhaps this:

```suggestion
This proposal is specifically to deal with the collection of bootables and improve our ability to treat the collection consistently.
```

# Goals
[goals]: #goals

- Enable a more uniform bootloader feature support across our packaged bootloaders.
  Concretely, making it relatively straightforward to convert most of the empty spaces in the feature matrix to YES's.
- Enable users of NixOS to implement custom bootloader tools and policy without needing to dive through the system profiles, and without patching Nixpkgs / NixOS.
- Define a stable specification of a generation's boot data which internal and external users can rely on.
  Changes to the specification should go through an RFC.

### Non-Goals
[non-goals]: #non-goals

- Rewriting the existing bootloader backends to actually fill out the feature matrix.
  The goal of this RFC is to make the feature development *easier*, not actually do it.
- Supporting SecureBoot.
  The authors of this RFC have done work in this regard, but this RFC is not about SecureBoot.

# Proposed Solution
[proposed-solution]: #proposed-solution

> - Each NixOS generation will have a bootspec (a JSON document) at `$out/boot.v1.json` containing all of the boot properties for that generation.

Reviewer 2: Given that you already require the document to be `json` and have a `"schemaVersion": 1,` field, perhaps this is redundant?

Author: I could imagine having multiple `boot.x.json` documents in the output to create very easy migrations between breaking changes, and the `scehmaVersion` in the document as a run-time verification of what exactly you're looking at. What do you think?

Reviewer 2: If the file name and `schemaVersion` differ, is it an assertion error?

Author: I don't know about an assertion error, but it should be considered an error of some sort. Probably better behavior would be to discard it and attempt to use a different boot.vN.json.

> - Each NixOS generation will have a bootspec (a JSON document) at `$out/boot.v1.json` containing all of the boot properties for that generation.

Author: Imagining a scenario where we have a v2 schema, we should probably still generate a v1 document. In this scenario we'd have a growing number of boot.vX.json documents. It'd probably be good to move these into a subdirectory.

```suggestion
- Each NixOS generation will have a bootspec (a JSON document) at `$out/bootspec/boot.v1.json` containing all of the boot properties for that generation.
```

Author: https://github.com/DeterminateSystems/bootspec-secureboot/pull/143 & https://github.com/NixOS/nixpkgs/pull/172237/commits/9b4df1d50f6df092cb195d630972ab36835d4f49

  "initrd": "/nix/store/xxx-initrd-linux/initrd",

  // Optional path to a tool to dynamically add secrets to an initrd
  "initrdSecrets": "/nix/store/xxx-append-secrets/bin/append-initrd-secrets",

  // Path to the kernel image
  "kernel": "/nix/store/xxx-linux/bzImage",

  // Kernel commandline options
  "kernelParams": [
    "amd_iommu=on",
    "amd_iommu=pt",
    "iommu=pt",
    "kvm.ignore_msrs=1",
    "kvm.report_ignored_msrs=0",
    "udev.log_priority=3",
    "systemd.unified_cgroup_hierarchy=1",
    "loglevel=4"
  ],

  // The label of the system. It should contain the operating system, kernel version,
  // and other user-relevant information to identify the system. This corresponds
  // loosely to `config.system.nixos.label`.
  "label": "NixOS 21.11.20210810.dirty (Linux 5.15.30)",

  // Top level path of the closure, in case some spelunking is required
  "toplevel": "/nix/store/xxx-nixos-system-xxx",

  "specialisation": {
    // <name> corresponds to <name> in specialisation.<name>.configuration.
    // Note: a specialisation's bootspec document should not contain any specialisations.
    "<name>": {
      // bootspec is the path to a bootspec document

>       "bootspec": "/path/to/a/bootspec.v1.json",

Reviewer 3: Where is this path expected to be? In the nix store? In the file directory?

I can imagine tools parsing this file outside a system with a viable nix-store.

Author: I'm inclined to move to embedding the full bootspec for each specialisation into the parent document, to address problems like this. I've written a patch to do that on my implementation PR, and will update the tools and RFC to match: https://github.com/NixOS/nixpkgs/pull/172237/commits/33fa3bd4409d47872ef88454eadad5950d5f691c

Reviewer 4: Embedding the full document is the way to go imo.

Author: I've also updated the bootspec datatypes and synthesis tool to match: https://github.com/DeterminateSystems/bootspec/pull/2 and corresponding secureboot implementation: https://github.com/DeterminateSystems/bootspec-secureboot/pull/141 I can update the RFC later today / tomorrow morning.

    }
  }
}
```

### Risks
[risks]: #risks

- Some of the bootloader backends are quite complicated, and in many cases have inadequate tests.

>   We could accidentally break corner cases.

Reviewer 5: Is it supposed that initial rewrite should produce exactly same menu? Should it be an option to compare two ways of generating things at activation time, so that users have easier time checking if their setup is properly supported after rewrite?

Author: It should produce a nearly identical menu, though minor details like the formatting of the information in the text may be different.

- The bootloader backends are inherently a weak point for NixOS, as it is our last option for rolling back.
  We cannot roll back a broken bootloader.

>   This and the previous point are risks, but also help demonstrate the value of reducing the amount of code and complexity in the generator.

Reviewer 6: Could 2 bootloader partitions be an option to mitigate this risk? An A/B style thing where someone uses an external bootloader and also one of our more tested options as a fallback like systemd-boot/grub

It wouldn't have to be default but could be opt in

Author: I don't think solving the problem like that is in scope for this RFC. There could be other work around making the bootloader A/B'd, but in the risks section for this RFC I think we should be focusing on ways to mitigate the risks of this specific RFC, not the systemic risk of a bad bootloader which NixOS already has.

Reviewer 5: Specific new risks (rewriting all the things) could probably be mitigated by at least encouraging support for before/after comparisons at the generate-the-menu-time. Hopefully the users with rare corner cases know they have rare corner cases and will use whatever consistency checks are offered…

- The bootloader backends will avoid reading data from the other files and directories when possible, preferring the information in the bootspec.
- A bootspec synthesizing tool will be used to synthesize a bootspec for generations which don't already have one.
  This tool will be shared across all of the bootloader backends, helping produce more uniform behavior.
- Existing bootloader backends will be updated to read properties from the bootspec, removing most if not all of their filesystem-spelunking code.

### Bootspec Format v1
[format-v1]: #format-v1

Using the following JSON:

```json5
{
  // Version of the specification used in the document
  "schemaVersion": 1,

  // Path to the stage-2 init, executed by the initrd
  "init": "/nix/store/xxx-nixos-system-xxx/init",

  // Path to the initrd
  "initrd": "/nix/store/xxx-initrd-linux/initrd",

  // Optional path to a tool to dynamically add secrets to an initrd
  "initrdSecrets": "/nix/store/xxx-append-secrets/bin/append-initrd-secrets",

  // Path to the kernel image
  "kernel": "/nix/store/xxx-linux/bzImage",

  // Kernel commandline options

>   "kernelParams": [

Reviewer 5: But kernel command line is also used to pass configuration to early init, so it is a bit less limited.

A list gives an option to do at least some filtering.

>   "kernelParams": [

Reviewer 4: There's no drawbacks to keeping this a list, while if it was squished in a string it would lose a bit of the meaning of the input. At least that's what I see here.

> A list gives an option to do at least some filtering.

It's probably too late to filter at this point; this is an input for tools that serialize the generation's boot configuration. But true, it wouldn't be as easy to filter otherwise.

>   "kernelParams": [

Reviewer 5: I assume that a) there will be unpredictable mix of data for kernel / PID1 / activation, b) the type choice here will propagate into some random place.

So a list costs little, allows asserting lack of spaces in entries if we want (pretty sure _something_ will be broken by them), and increases chances to filter out a single specific option before bootspec being written. Overall, a win, in my opinion.

  // Top level path of the closure, in case some spelunking is required
  "toplevel": "/nix/store/xxx-nixos-system-xxx",

  "specialisation": {
    // <name> corresponds to <name> in specialisation.<name>.configuration.
    // Note: a specialisation's bootspec document should not contain any specialisations.
    "<name>": {
      // A full Bootspec v1 document.
      // Note that it is not valid for a specialisation to have further specialisations.
    }
  }
}
```

### Risks
[risks]: #risks

- Some of the bootloader backends are quite complicated, and in many cases have inadequate tests.
  We could accidentally break corner cases.
- The bootloader backends are inherently a weak point for NixOS, as it is our last option for rolling back.
  We cannot roll back a broken bootloader.
  This and the previous point are risks, but also help demonstrate the value of reducing the amount of code and complexity in the generator.

### Milestones
[milestones]: #milestones

- Create and package the backwards compatibility synthesizer as a standalone tool.
  A version of it already exists, but it is not standalone.
- Generate the bootspec files as part of building the system closure.

> - Update the bootloader backends to use bootspec as its primary source of installation data.

Reviewer 7: nit
```suggestion
- Update the bootloader backends to use bootspec as their primary source of installation data.
```

---
feature: bootspec
start-date: 2022-05-09
author: Graham Christensen
co-authors: Cole Helbling

> shepherd-team: (names, to be nominated and accepted by RFC steering committee)
> shepherd-leader: (name to be appointed by RFC steering committee)

Reviewer 8: ```suggestion
shepherd-team: @lheckemann, @06kellyjac, @GovanifY, @RaitoBezarius
shepherd-leader: @lheckemann
```

  A bootspec document named `boot.json` must have a `"v1"` key that contains the body of the specification.
  Generally, the key must follow the format of `v${schemaVersion}`.
    - The bootloader installation phase is relatively unchanged from the way it is now.
      The bootloader backend will have an executable that is run against a collection of generations, and the backend is any of the currently supported backends plus an "external" backend which the user can define.
- The bootloader backends will avoid reading data from the other files and directories when possible, preferring the information in the bootspec.
- A bootspec synthesizing tool will be used to synthesize a bootspec for generations which don't already have one.
  This tool will be shared across all of the bootloader backends, helping produce more uniform behavior.
- Existing bootloader backends will be updated to read properties from the bootspec, removing most if not all of their filesystem-spelunking code.

### Bootspec Format v1
[format-v1]: #format-v1

Using the following JSON:

```json5
{
  // Toplevel key describing the version of the specification used in the document
  "v1": {

> 

Reviewer 9: Don't we need to know the target system architecture in the bootspec?

Author: I don't know that you would, but perhaps we would. What for?

Reviewer 3: I don't think the arch should be included in the present document.

I could see a 'universal media' in the future where the bootloader can filter the derivations compatible with the system architecture, but that seems too far-fetched to be included in v1.

Author: I think we should add the architecture.

Reviewer 10: What was mentioned by @lheckemann during the meeting was (paraphrased): putting the arch helps to do the decision of "do I need BootX64 or BootX32, etc. artifacts?" IIRC. (correct me if I'm wrong)

    // Path to the stage-2 init, executed by the initrd
    "init": "/nix/store/xxx-nixos-system-xxx/init",

    // Path to the initrd
    "initrd": "/nix/store/xxx-initrd-linux/initrd",

    // Optional path to a tool to dynamically add secrets to an initrd
    "initrdSecrets": "/nix/store/xxx-append-secrets/bin/append-initrd-secrets",

    // Path to the kernel image
    "kernel": "/nix/store/xxx-linux/bzImage",

    // Kernel commandline options
    "kernelParams": [
      "amd_iommu=on",
      "amd_iommu=pt",
      "iommu=pt",
      "kvm.ignore_msrs=1",
      "kvm.report_ignored_msrs=0",
      "udev.log_priority=3",
      "systemd.unified_cgroup_hierarchy=1",
      "loglevel=4"
    ],

    // The label of the system. It should contain the operating system, kernel version,
    // and other user-relevant information to identify the system. This corresponds
    // loosely to `config.system.nixos.label`.

>     "label": "NixOS 21.11.20210810.dirty (Linux 5.15.30)",

Reviewer 11: Some bootloaders (GRUB for instance) also record the date when the generation was made (probably mtime of the generation symlink?). Would that be included in the label or a separate field?

Author: The tooling / automation is going to have to identify a collection of bootspec documents to use. It could be a directory of symlinks, or a database, or a tool run in CI -- but wherever it comes from should probably provide this sort of information.

Reviewer 12: FWIW, I've implemented the current logic (well, an amalgamation of the current logic from different scripts) as a reusable component here: https://gitlab.com/K900/bootis/-/tree/main/nixos-list-generations

The main problem with that approach is there are a lot of files that you need to look for and examine in specific ways.
We end up duplicating a lot of work across bootloader backends to "learn" about the system.
By creating a source of truth for this information the bootloader backends can focus more on the work that is unique to their problem.

Doing this work now may also make it easier to refactor a NixOS system's `$out` , reducing the number of symlinks and "noise" if we wanted to do that.

**Does this interrupt the work to make stage-1 based on systemd?**

No.
Stage-1 won't use this document at all.

**What about memtest and other boot options?**

That should be left to configuration passed to the bootloader backend.
There could be any number of "extra" bootables like this, and importantly they have nothing to do with a specific NixOS system closure.
Concretely, a user who enables memtest probably wants the most recent memtest, and not one memtest from every generation.

# Open Questions
[open-questions]: #open-questions

- Should there be a general-purpose "meta" or "extensions" field which allows arbitrary extension of the data? What should this look like?

> 

Reviewer 13: ```suggestion

- Should there be a non-derivation metadata facility that carries useful metadata for which it isn't desirable to cause rebuilds, redeployments or new boot entries - specifically data such a git commit hash or a date of last modification.

```

Reviewer 10: @roberth Could you clarify what you envision by a "non derivation metadata facility"? Would that be a generalization of bootspec so that bootspec would be only a special case of it? (my interpretation)

We discussed it during our meeting, but I think we had some different interpretations of your comments.

Reviewer 13: By adding a field such as the git revision or last modified date, you're creating new generations that are otherwise identical to preceding ones when you make changes to unrelated files. My open question aims to prevent that from happening.
(The need for this came up in https://github.com/NixOS/nixpkgs/pull/179772)

> Would that be a generalization of bootspec so that bootspec would be only a special case of it?

Possibly.

> > non-derivation

I suppose the metadata _can_ be in a derivation, but this derivation should not be used for determining the equivalence of generations. Assuming we keep using the output path of `system.build.toplevel` for the purpose of generation equivalence, the metadata could be put in a separate derivation that isn't part of `system.build.toplevel`.
There's a number of ways you could implement this sort of thing.

Making `toplevel` not top level isn't great. I could rephrase this as

```suggestion

- How should we identify generations as being equivalent? This will help keep boot menus concise and allows activation to be cut short, especially on systems with frequently changing metadata, such as the revision of a repository that contains components not deployed to the system.
```

This allows other solutions such as a "fingerprint" field containing a hash of the essential values so that duplicate generations can be detected by the bootloader menu generator (or other tool that benefits from a "shortcutting" behavior).

Implementation idea (not necessarily part of bootspec; more of a NixOS implementation detail. `fingerprint` existence and type (string) _is_ bootspec though.)
```
let spec0 = config.system.build.bootspec; # assuming this is where NixOS puts the almost-final bootspec stuff
     specHash = builtins.hashString (builtins.toJSON (config.system.build.bootspec // { meta = null; }));
     spec1 = spec0 // { fingerprint = specHash; };
in {
  # assuming somethign like https://github.com/NixOS/nixpkgs/pull/148456/commits/4ec415cff9bcbaef08e0d900406a5c7181e71881
  config.system.systemBuilderCommands = ''
    ln -s ${json.generate "bootspec.json" spec1} $out/bootspec.json
  '';
}
```

Bootloader backends, by definition reconfigure the boot partition and records, and are supposed to apply retroactively to old configurations.
This regeneration happens way beyond the lifecycle of the module system.

**How is this easier than `if (-f "$out/append-initrd-secrets")`?**

The main problem with that approach is there are a lot of files that you need to look for and examine in specific ways.
We end up duplicating a lot of work across bootloader backends to "learn" about the system.
By creating a source of truth for this information the bootloader backends can focus more on the work that is unique to their problem.

Doing this work now may also make it easier to refactor a NixOS system's `$out` , reducing the number of symlinks and "noise" if we wanted to do that.

**Does this interrupt the work to make stage-1 based on systemd?**

No.
Stage-1 won't use this document at all.

**What about memtest and other boot options?**

That should be left to configuration passed to the bootloader backend.
There could be any number of "extra" bootables like this, and importantly they have nothing to do with a specific NixOS system closure.

> Concretely, a user who enables memtest probably wants the most recent memtest, and not one memtest from every generation.

Reviewer 14: Another option (just an idea that came to my mind) would be to allow special "generation-less bootspec files" that have alternative fields. Instead of `kernel`, `kernelParams`, `initrd`, `init`, `system`, these files would simply have an `efi` field pointing to a valid EFI executable file.

That would allow a unified view of "bootable things" as an input to the bootloader backend: Either a generation (a directory containing a generation bootspec file) or a standalone bootable (a directory containing a standalone bootspec file).

  The goal of this RFC is to make the feature development *easier*, not actually do it.
- Supporting SecureBoot.
  The authors of this RFC have done work in this regard, but this RFC is not about SecureBoot.

> - Integrating Bootspec into an existing bootloader backend is not expected to perfectly produce the exact same result. For example, it is likely that the text describing the entrties in the menu may differ.

Reviewer 15: ```suggestion
- Integrating Bootspec into an existing bootloader backend is not expected to perfectly produce the exact same result. For example, it is likely that the text describing the entries in the menu may differ.
```
